### PR TITLE
UI/manage event

### PR DESCRIPTION
**Pull request by Keith-Go**
## Ticket
- [UI manage event](https://trello.com/c/0EPGiWyn/30-ui-event-manegement-page)

## To do
- List event which user create
- Show event detail manegement page
- Edit event page

## Video

https://github.com/luke-nguyen-goldenowl/join_me/assets/152830147/aae97487-9710-4c48-a7cc-e3ea1e8ab0c1

### CODE REVIEW

File: lib/src/features/add_event/logic/add_event_bloc.dart [hunk 1]
```diff
@@ -4,13 +4,18 @@ import 'package:flutter_map/flutter_map.dart';
...
 import 'package:myapp/src/features/add_event/logic/add_event_state.dart';
+import 'package:myapp/src/network/model/event/event.dart';
 
 class AddEventBloc extends Cubit<AddEventState> {
   AddEventBloc() : super(AddEventState.ds());
 
   PageController controller = PageController(initialPage: 0);
   MapController mapController = MapController();
 
+  void initState(MEvent event) {
```
Comment (luke-nguyen-goldenowl) [inline lib/src/features/add_event/logic/add_event_bloc.dart]: Is this method used for anything?

File: lib/src/features/detail_event/data/list_image.dart [hunk 2]
```diff
@@ -1,8 +1,6 @@
-import 'package:flutter/material.dart';
-import 'package:myapp/gen/assets.gen.dart';
-
-List<Image> listImage = [
-  Assets.images.images.landscape2.image(fit: BoxFit.cover),
-  Assets.images.images.landscape3.image(fit: BoxFit.cover),
-  Assets.images.images.landscape4.image(fit: BoxFit.cover)
+List<String> listImage = [
```
Comment (luke-nguyen-goldenowl) [inline lib/src/features/detail_event/data/list_image.dart]: Why not add it to mock repository?

File: lib/src/features/detail_event/view/detail_event_view.dart [hunk 3]
```diff
@@ -38,7 +39,13 @@ class DetailEventPage extends StatelessWidget {
                 Expanded(
                   child: CustomScrollView(
                     slivers: [
-                      const SliverAppBarCustomDetailEvent(),
+                      SliverAppBarCustomDetailEvent(
+                        indexPageImage: state.indexPageImage,
+                        images: listImage,
```
Comment (luke-nguyen-goldenowl) [inline lib/src/features/detail_event/view/detail_event_view.dart]: After adding "listImage" to the mock repository, create a function to call the API to get event details using the id passed from the constructor, handle in bloc and state, remember to update buildWhen

File: lib/src/features/manage_event/edit_event/logic/edit_event_bloc.dart [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+import 'package:myapp/src/features/manage_event/edit_event/logic/edit_event_state.dart';
+import 'package:myapp/src/network/model/event/event.dart';
+
+class EditEventBloc extends Cubit<EditEventState> {
+  EditEventBloc() : super(EditEventState.ds());
+
+  PageController controller = PageController(initialPage: 0);
+  MapController mapController = MapController();
+
+  void initState(MEvent event) {
```
Comment (luke-nguyen-goldenowl) [inline lib/src/features/manage_event/edit_event/logic/edit_event_bloc.dart:13]: Just using event param is enough, No need to create other params in the state

File: lib/src/features/manage_event/edit_event/widget/address_page_edit_event.dart [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+import 'package:myapp/src/features/manage_event/edit_event/logic/edit_event_bloc.dart';
+import 'package:myapp/src/features/manage_event/edit_event/logic/edit_event_state.dart';
+import 'package:myapp/src/theme/colors.dart';
+
+class AddressPageEditEvent extends StatelessWidget {
+  const AddressPageEditEvent({super.key});
+
+  @override
+  Widget build(BuildContext context) {
+    return BlocBuilder<EditEventBloc, EditEventState>(
```
Comment (luke-nguyen-goldenowl) [inline lib/src/features/manage_event/edit_event/widget/address_page_edit_event.dart:14]: add buildWhen

File: lib/src/features/manage_event/edit_event/view/edit_event_view.dart [hunk 6]
```diff
@@ -1,18 +1,142 @@
...
+
+class EditEventPage extends StatelessWidget {
+  const EditEventPage({
+    super.key,
+  });
+
+  @override
+  Widget build(BuildContext context) {
+    Size size = MediaQuery.of(context).size;
+    return BlocBuilder<EditEventBloc, EditEventState>(
```
Comment (luke-nguyen-goldenowl) [inline lib/src/features/manage_event/edit_event/view/edit_event_view.dart:32]: update buildWhen

File: lib/src/features/manage_event/edit_event/widget/detail_page_edit_event.dart [hunk 7]
```diff
@@ -0,0 +1,133 @@
...
+
+import 'package:myapp/widgets/forms/input.dart';
+
+class DetailPageEditEvent extends StatelessWidget {
+  const DetailPageEditEvent({super.key});
+
+  @override
+  Widget build(BuildContext context) {
+    final EditEventBloc editEventBloc = BlocProvider.of<EditEventBloc>(context);
+    return BlocBuilder<EditEventBloc, EditEventState>(
```
Comment (luke-nguyen-goldenowl) [inline lib/src/features/manage_event/edit_event/widget/detail_page_edit_event.dart:16]: add buildWhen

File: lib/src/features/manage_event/logic/manage_event_bloc.dart [hunk 8]
```diff
@@ -0,0 +1,22 @@
+import 'package:flutter_bloc/flutter_bloc.dart';
+import 'package:myapp/src/features/manage_event/logic/manage_event_state.dart';
+import 'package:myapp/src/network/domain_manager.dart';
+import 'package:myapp/src/network/model/common/pagination/pagination.dart';
+import 'package:myapp/src/network/model/event/event.dart';
+
+class ManageEventBloc extends Cubit<ManageEventState> {
```
Comment (luke-nguyen-goldenowl) [inline lib/src/features/manage_event/logic/manage_event_bloc.dart]: Update with pagination bloc

File: lib/src/features/manage_event/logic/manage_event_state.dart [hunk 9]
```diff
@@ -0,0 +1,32 @@
+// ignore_for_file: public_member_api_docs, sort_constructors_first
+import 'package:myapp/src/network/model/common/pagination/pagination.dart';
+import 'package:myapp/src/network/model/event/event.dart';
+
+class ManageEventState {
```
Comment (luke-nguyen-goldenowl) [inline lib/src/features/manage_event/logic/manage_event_state.dart]: pagination state

File: lib/src/features/manage_event/manage_event_detail/view/manage_event_detail_view.dart [hunk 10]
```diff
@@ -1,26 +1,162 @@
...
+
+class ManageEventDetailPage extends StatelessWidget {
+  const ManageEventDetailPage({super.key, required this.id});
+
+  final String id;
+
+  @override
+  Widget build(BuildContext context) {
+    return BlocBuilder<ManageEventDetailBloc, ManageEventDetailState>(
+      buildWhen: (previous, current) => previous != current,
```
Comment (luke-nguyen-goldenowl) [inline lib/src/features/manage_event/manage_event_detail/view/manage_event_detail_view.dart]: update buildWhen

File: lib/src/features/manage_event/manage_event_detail/view/manage_event_detail_view.dart [hunk 11]
```diff
@@ -1,26 +1,162 @@
...
+      builder: ((context, state) {
+        return Scaffold(
+          backgroundColor: AppColors.white,
+          body: Column(
```
Comment (luke-nguyen-goldenowl) [inline lib/src/features/manage_event/manage_event_detail/view/manage_event_detail_view.dart]: You should separate the witgets to make it easier to debug and handle buildwhen

File: lib/src/features/manage_event/view/manage_event_view.dart [hunk 12]
```diff
@@ -1,26 +1,49 @@
...
+            appBar: const AppBarCustom(
+              title: "My Events",
+            ),
+            body: Column(
+              children: [
+                Expanded(
+                  child: ListView.builder(
+                    itemCount: state.pagination.data.length + 1,
+                    itemBuilder: ((context, index) {
+                      return index == state.pagination.data.length
```
Comment (luke-nguyen-goldenowl) [inline lib/src/features/manage_event/view/manage_event_view.dart]: We should use if instead of "?" and ":" if possible

**Comment by Keith-Go** [comment]
resolved

File: lib/src/features/detail_event/view/detail_event_view.dart [hunk 13]
```diff
@@ -41,7 +40,7 @@ class DetailEventPage extends StatelessWidget {
                     slivers: [
                       SliverAppBarCustomDetailEvent(
                         indexPageImage: state.indexPageImage,
-                        images: listImage,
+                        images: state.event!.images,
```
Comment (luke-nguyen-goldenowl) [inline lib/src/features/detail_event/view/detail_event_view.dart]: Don't use "!"

File: lib/src/features/manage_event/edit_event/widget/address_page_edit_event.dart [hunk 14]
```diff
@@ -12,59 +12,62 @@ class AddressPageEditEvent extends StatelessWidget {
...
-                          },
+                ),
+                const SizedBox(height: 10),
+                if (state.event.location != null)
+                  Expanded(
+                    child: FlutterMap(
+                      mapController:
+                          context.read<EditEventBloc>().mapController,
+                      options: MapOptions(
+                        center: state.event.location!,
```
Comment (luke-nguyen-goldenowl) [inline lib/src/features/manage_event/edit_event/widget/address_page_edit_event.dart:36]: Don't use "!"

File: lib/src/features/detail_event/view/detail_event_view.dart [hunk 15]
```diff
@@ -41,7 +41,7 @@ class DetailEventPage extends StatelessWidget {
                     slivers: [
                       SliverAppBarCustomDetailEvent(
                         indexPageImage: state.indexPageImage,
-                        images: state.event!.images,
+                        images: state.event!.images!,
```
Comment (luke-nguyen-goldenowl) [inline lib/src/features/detail_event/view/detail_event_view.dart]: don't use "!"

File: lib/src/features/manage_event/edit_event/widget/detail_page_edit_event.dart [hunk 16]
```diff
@@ -32,7 +32,7 @@ class DetailPageEditEvent extends StatelessWidget {
                     children: [
                       XInput(
                         key: const Key('addEvent_nameEventInput_textField'),
-                        value: state.event.name,
+                        value: state.event.name!,
```
Comment (luke-nguyen-goldenowl) [inline lib/src/features/manage_event/edit_event/widget/detail_page_edit_event.dart]: don't use "!"

File: lib/src/features/manage_event/edit_event/widget/detail_page_edit_event.dart [hunk 17]
```diff
@@ -45,7 +45,7 @@ class DetailPageEditEvent extends StatelessWidget {
                             'addEvent_descriptionEventInput_textField'),
                         minLines: 1,
                         maxLines: 10,
-                        value: state.event.description,
+                        value: state.event.description!,
```
Comment (luke-nguyen-goldenowl) [inline lib/src/features/manage_event/edit_event/widget/detail_page_edit_event.dart]: don't use "!"

File: lib/src/features/manage_event/widget/manage_event_item.dart [hunk 18]
```diff
@@ -13,7 +13,7 @@ class ManageEventItem extends StatelessWidget {
   Widget build(BuildContext context) {
     return GestureDetector(
       onTap: () {
-        AppCoordinator.showManageEventDetails(id: event.id);
+        AppCoordinator.showManageEventDetails(id: event.id!);
```
Comment (luke-nguyen-goldenowl) [inline lib/src/features/manage_event/widget/manage_event_item.dart]: don't use "!"

File: lib/src/features/manage_event/widget/manage_event_item.dart [hunk 19]
```diff
@@ -35,7 +35,7 @@ class ManageEventItem extends StatelessWidget {
             ClipRRect(
               borderRadius: BorderRadius.circular(12),
               child: Image.asset(
-                event.images[0],
+                event.images![0],
```
Comment (luke-nguyen-goldenowl) [inline lib/src/features/manage_event/widget/manage_event_item.dart]: don't use "!"

File: lib/src/features/manage_event/widget/manage_event_item.dart [hunk 20]
```diff
@@ -48,7 +48,7 @@ class ManageEventItem extends StatelessWidget {
                 crossAxisAlignment: CrossAxisAlignment.start,
                 children: [
                   Text(
-                    event.name,
+                    event.name!,
```
Comment (luke-nguyen-goldenowl) [inline lib/src/features/manage_event/widget/manage_event_item.dart]: don't use "!"

File: lib/src/network/data/event/event_repository_mock.dart [hunk 21]
```diff
@@ -74,7 +74,7 @@ class EventRepositoryMock {
     if (search == "") return MResult.success([]);
     final result = listEvent
         .where((element) =>
-            element.name.toLowerCase().contains(search.toLowerCase()))
+            element.name!.toLowerCase().contains(search.toLowerCase()))
```
Comment (luke-nguyen-goldenowl) [inline lib/src/network/data/event/event_repository_mock.dart]: don't use "!"

File: lib/src/features/manage_event/edit_event/logic/edit_event_bloc.dart [hunk 22]
```diff
@@ -0,0 +1,59 @@
...
+    emit(state.copyWith(
+      event: event,
+      // selectedLocation: event.location,
```
Comment (luke-nguyen-goldenowl) [inline lib/src/features/manage_event/edit_event/logic/edit_event_bloc.dart]: Remove unuse comments

File: lib/src/features/manage_event/edit_event/logic/edit_event_state.dart [hunk 23]
```diff
@@ -0,0 +1,57 @@
...
+    return EditEventState(
+      currentPage: 0,
+      event: MEvent(id: '1', host: MUser.empty()),
+      time: null,
+    );
+  }
+
+  EditEventState copyWith({
+    MEvent? event,
+    LatLng? selectedLocation,
```
Comment (luke-nguyen-goldenowl) [inline lib/src/features/manage_event/edit_event/logic/edit_event_state.dart]: remove unuse params

**Comment by Keith-Go** [comment]
resolved

**Comment by luke-nguyen-goldenowl** [comment]
Approved

**Pull request merged**